### PR TITLE
[Client] Add deprecation warnings for direct ray.client().connect() calls

### LINKED ISSUE #17508: [client] Add warnings when ClientBuilder is used directly instead of through ray.init(ray://)
### Describe your feature request

Warning message should include the exact ray.init call needed to replace the old ray.client(...) client call. 

For questions, post on our Discussion page instead: https://discuss.ray.io/

### PR DESCRIPTION
**Pull request by ckw017**
## Why are these changes needed?

`ray.client().connect()` has been deprecated in favor of `ray.init`. These changes will generate a warning when `ray.client()` is used with the appropriate `ray.init` command to replace it.
 
Screen Shot 2021-09-21 at 9 52 03 AM

## Related issue number

Closes #17508

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by AmeerHajAli** [comment]
Is this ready for review?

File: python/ray/tests/test_client_builder.py [hunk 1]
```diff
@@ -283,5 +284,98 @@ def test_address_resolution(call_ray_stop_only):
...
 
 
+def mock_connect(*args, **kwargs):
+    """
+    Force exit instead of actually attempting to connect
+    """
+    raise ConnectionError
+
+
+def has_client_deprecation_warn(warning, expected_replacement):
```
Comment (ijrsvt) [inline python/ray/tests/test_client_builder.py:294]: Could you add type hints for this?

File: python/ray/tests/test_client_builder.py [hunk 2]
```diff
@@ -283,5 +284,98 @@ def test_address_resolution(call_ray_stop_only):
             del os.environ["RAY_ADDRESS"]
 
 
+def mock_connect(*args, **kwargs):
+    """
+    Force exit instead of actually attempting to connect
+    """
+    raise ConnectionError
```
Comment (ijrsvt) [inline python/ray/tests/test_client_builder.py:291]: Could we maybe try to pass out a mock to actually have the connect succeed. I just want to make sure that this warning does not stop the connection.

File: python/ray/tests/test_client_builder.py [hunk 3]
```diff
@@ -283,5 +284,98 @@ def test_address_resolution(call_ray_stop_only):
...
+    if expected_replacement not in message:
+        return False
+    return True
+
+
+def test_client_deprecation_warn():
+    """
+    Tests that calling ray.client directly raises a deprecation warning with
+    a copy pasteable replacement for the client().connect() call converted
+    to ray.init style.
```
Comment (ijrsvt) [inline python/ray/tests/test_client_builder.py:308]: Can you also add some tests for this with environment variables as well? (i.e. that they are not printed in the warnings)?

**Review by ijrsvt — COMMENTED** [review]
Looks quite good, a few minor comments!

**Review by AmeerHajAli — APPROVED** [review]
LGTM. a few minor comments.

**Comment by ckw017** [comment]
> BTW, how is this going to work in the product? will the user get any warnings?
At the moment the user won't get warnings, the product client builder will need a separate update. Better off this way since product might be running with a version of ray that doesn't support ray.init

**Pull request merged**